### PR TITLE
fix(nats): Client-Free(ish) NATS container

### PR DESCRIPTION
**Pull request by bearrito**
@totallyzen 
Closed previous PR because of dirty history and bad-merge.

Based of this comment - https://github.com/testcontainers/testcontainers-python/issues/446#issuecomment-1986860683 
I removed hard dependency on the NATS client.

This PR does not require a client to work or validate at a basic level.
However if a developer wants to use a client, there is test support for that.

### CODE REVIEW

File: modules/nats/tests/test_nats.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+If you are developing this and you want to test more advanced scenarios using a client
+Activate your poetry shell.
+pip install nats-py
+This will get nats-py into your environment but keep it out of the project
+
+
+"""
+
+
+NO_NATS_CLIENT = True
```
Comment (bearrito) [inline modules/nats/tests/test_nats.py:16]: This section will check if you have the client available, if so then certain tests are available. If not then those tests are skipped.

File: modules/nats/testcontainers/nats/__init__.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
+#
```
Comment (bearrito) [inline modules/nats/testcontainers/nats/__init__.py:1]: No client imported in this file.

File: modules/nats/tests/test_nats.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    Returns:
+        client: Nats client to connect to the container.
+    """
+    conn_string = container.nats_uri()
+    client = await nats_connect(conn_string)
+    return client
+
+
+def test_basic_container_ops():
+    with NatsContainer() as container:
```
Comment (bearrito) [inline modules/nats/tests/test_nats.py:39]: This tests runs no matter what.

File: modules/nats/tests/test_nats.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        uri = container.nats_uri()
+        management_uri = container.nats_management_uri()
+
+        assert uri != management_uri
+
+
+pytest.mark.usefixtures("anyio_backend")
+
+
+@pytest.mark.skipif(NO_NATS_CLIENT, reason="No NATS Client Available")
```
Comment (bearrito) [inline modules/nats/tests/test_nats.py:53]: Skipped if no client.

**Comment by alexanderankin** [comment]
@bearrito for what its worth we havent decided what we are doing with the dependencies yet - i was going to merge your PR as is after releasing whatever other fixes we can fit before we add features (things that have to bump minor version)

File: modules/nats/tests/test_nats.py (hunk 1 above)
Comment (alexanderankin) [reply modules/nats/tests/test_nats.py:16]: there's really no reason to exclude tests due to missing deps - just dont force a dep on the client code which is using the testcontainer

**Comment by bearrito** [comment]
@alexanderankin No worries. If you guys decide to allow more dependencies. I can add to the extras and then remove the try/catch import

File: modules/nats/tests/test_nats.py (hunk 1 above)
Comment (bearrito) [reply modules/nats/tests/test_nats.py:16]: What's the correct way to indicate that with poetry?

**Comment by totallyzen** [comment]
@bearrito my take at this point is that you can add it as a dev dependency - the point is more so to prevent enforcement of clients, not to make testing a hellscape.

I've reflected this in the latest changes on #460 as well, comment there instead on this if you agree!

**Comment by alexanderankin** [comment]
what happened here? i am rebasing all the branches because too crazy otherwise

**Comment by bearrito** [comment]
@alexanderankin  I think I was in the middle of you doing that. What do you want me to do here?

**Comment by alexanderankin** [comment]
@bearrito no problem, i just merged all the newer ones, about to rebase this one and merge it. thanks for your contribution and support here. let me know if you want to have a final look over in a few mins before i push and hit merge.

**Comment by alexanderankin** [comment]
need to fix the doctest still, one min

File: modules/nats/tests/test_nats.py (hunk 1 above)
Comment (alexanderankin) [reply modules/nats/tests/test_nats.py]: > What's the correct way to indicate that with poetry?

extremely painfull :smile:

**Pull request merged**